djtriggers/logic.py

```diff
@@ -147,3 +102,58 @@ def clean_triggers(expiration_dt=None, type_to_table=None):
         nr_deleted += 1
 
     return nr_deleted
+
+
+def _process_async(triggers, use_statsd):
+    """
+    Process given triggers asynchronously
+    :param django.db.models.query.QuerySet triggers: triggers to be processed
+    :param bool use_statsd: whether to use statsd
+    :return: None
+    """
+    logger.info('Start processing %d triggers of type %s asynchronously', triggers.count(), triggers[0].trigger_type)
+    # create subtasks for all triggers to be processed.
+    # 'max_retries' defines the maximum number of retries that can be performed by Celery.
+    # This is done to manually disable retries at Celery workers
+    tasks = [trigger.process.subtask({'use_statsd': use_statsd},
+                                     max_retries=getattr(settings, 'DJTRIGGERS_CELERY_TASK_MAX_RETRIES', 0))
+             for trigger in triggers]
+    # Execute all subtasks asynchronously
+    group(tasks).apply_async()
+
+
+def _process_sync(triggers, use_statsd):
+    """
+    Process given triggers synchronously
+    :param django.db.models.query.QuerySet triggers: triggers to be processed
+    :param bool use_statsd: whether to use statsd
+    :return: None
+    """
+    logger.info('Start processing %d triggers of type %s synchronously', triggers.count(), triggers[0].trigger_type)
+
+    count_done, count_error, count_exception = 0, 0, 0
+
+    # Process each trigger
+    for trigger in triggers:
+        try:
+            trigger.process.apply(use_statsd=use_statsd)
+        # The trigger didn't need processing yet
+        except (ProcessLaterError, AlreadyProcessedError):
+            pass
+        # The trigger raised an (expected) error while processing
```

Comment:
Why is there a difference between ProcessError exception and others? What kind of errors are "expected" then?